auditing_standards.md

```diff
@@ -237,3 +237,76 @@ Crates with this criteria contain very dangerous unsafe rust code.
     *   Undefined behavior may be possible under common circumstances.
 *   Most crates that try to be sound but don't quite make the cut go in `ub-risk-3`, not here. These crates are wildly unsafe, and the only time we should use them is when they are a necessary evil.
 *   Everything worse than `ub-risk-3` goes in here and we should do our best to avoid using them.
+
+## Criteria-agnostic guidelines
+
+### Delta audits should describe the final version
```

Comment:
This change assumes that doing this:

1. Audit crate version 1 with ub-risk-3.
2. Audit crate delta 1->2 with ub-risk-0.

Results in crate version 2 having ub-risk-0. To my mind, this series of events should still result in version 2 having ub-risk-3, but I may understand how the audits are interacting. As I understand it, delta audits can only ever remove audited criteria by [acting as non-traversable edges](https://github.com/mozilla/cargo-vet/blob/main/src/resolver.rs#L1598-L1601). If that is the case, then we should **not** require delta audits to describe the final state of the crate.

Reply:
I believe upstream documents delta audits as being basically the New audit

Reply:
> This change assumes that doing this:
> 
> 1. Audit crate version 1 with ub-risk-3.
> 2. Audit crate delta 1->2 with ub-risk-0.
> 
> Results in crate version 2 having ub-risk-0. To my mind, this series of events should still result in version 2 having ub-risk-3

Yes, your understanding is correct.
* Version 1 is `ub-risk-3` (because of direct audit) and `ub-risk-4` (because `ub-risk-3` `implies` `ub-risk-4`)
* Edge between version 1 and version 2 delta is `ub-risk-0` (because of direct audit) and `ub-risk-1`, `ub-risk-2`, `ub-risk-3`, `ub-risk-4` (because `ub-risk-N` `implies` `ub-risk-<N+1>`)
* Version 2 is *not* `ub-risk-0` because version 1 was not `ub-risk-0` and therefore the edge doesn't apply

But... I think that:
* Having explicit guidance what to do would be desirable
* I think that stating the final state is more useful, because:
    - Saying `ub-risk-0` is misleading when the crate has `unsafe`.  In the future, when I review an audit of a delta I want to know if there is `unsafe` (so that I can evaluate if changes of safe code may impact crate-internal invariants of unchanged `unsafe` code elsewhere in the crate).  And tracing back through delta audit edges seems a bit painful.  I guess I could also patch-in each audit and grep for `unsafe`, but this is also a bit of extra work.
    - In general, a delta review of `ub-risk-2` has to consider the overall crate (and not just the delta) and therefore certifying the final state seems more correct.

Reply:
https://mozilla.github.io/cargo-vet/audit-entries.html#delta

Reply:
@djkoloski we've been pretty good about this in Google3: we will often just review a diff and apply a "diff review only" tag that doesn't get upstreamed as an audit. We mark things as full audits when we have looked at past audits, which is pretty easy given that we make sure we track comments. Or just look at the rest of the unsafe.

I don't really see a question of us changing this norm, given that this is in cargo vet's docs:

> The standard here is that the properties are actually preserved, not merely that that the diff doesn't obviously violate them. It is the responsibility of the auditor to acquire sufficient context to certify the former.

We'd be violating that contract if we changed this.

Reply:
Alright, I'm fine with this line of reasoning.